### PR TITLE
GH-89727: Fix `os.fwalk()` recursion error on deep trees

### PR DESCRIPTION
**Pull request by barneygale**
Implement `os.fwalk()` using a list as a stack to avoid emitting recursion errors on deeply nested trees.

* Issue: gh-89727

### CODE REVIEW

**Comment by barneygale** [comment]
https://github.com/python/cpython/pull/119473 fixes the test failure.

File: Lib/os.py [hunk 1]
```diff
@@ -528,17 +539,11 @@ def _fwalk(topfd, toppath, isbytes, topdown, onerror, follow_symlinks):
             except OSError as err:
                 if onerror is not None:
                     onerror(err)
-                continue
-            try:
+            else:
+                stack.append((None, dirfd))
```
Comment (barneygale) [inline Lib/os.py]: Need to defer the `open()` too otherwise we consume tons of FDs.

**Comment by barneygale** [comment]
Ready for review! I fixed the test failure mentioned above by reversing the `dirs` and `entries`, though IMHO this shouldn't be necessary in bottom-up mode. The diff is fairly large because a lot of work was happening after the recursive call in the previous version, and all that code needed to be brought up front. Fortunately that allows for some de-duplication of `open()` and `stat()` calls.

File: Lib/os.py [hunk 2]
```diff
@@ -514,31 +542,18 @@ def _fwalk(topfd, toppath, isbytes, topdown, onerror, follow_symlinks):
...
-                if follow_symlinks or path.samestat(orig_st, stat(dirfd)):
-                    dirpath = path.join(toppath, name)
-                    yield from _fwalk(dirfd, dirpath, isbytes,
-                                      topdown, onerror, follow_symlinks)
-            finally:
-                close(dirfd)
-
-        if not topdown:
-            yield toppath, dirs, nondirs, topfd
+        toppath = path.join(toppath, toppath[:0])  # Add trailing slash.
```
Comment (carljm) [inline Lib/os.py:555]: I guess we use `toppath[:0]` here as a clever alternative to `b"" if isbytes else ""`?

**Review by carljm — APPROVED** [review]
I won't claim I did a line-by-line verification of equivalence with the current code, but I walked through the new code and it looks sensible, and the existing test coverage looks reasonable.

Did you consider adding a test that sets a very low recursion limit and verifies we don't hit it? I don't feel strongly about this.

File: Lib/os.py (hunk 2 above)
Comment (barneygale) [reply Lib/os.py:555]: That's right. Do you think it's needlessly clever? If you prefer your version I'm happy to change.

The `[:0]` thing is used a few times in posixpath.py and ntpath.py, FWIW.

**Comment by barneygale** [comment]
> Did you consider adding a test that sets a very low recursion limit and verifies we don't hit it? I don't feel strongly about this.

I've enabled `test_walk_above_recursion_limit` for `fwalk()`, which tests with the recursion limit set to 45 and a directory depth of 50. Previously that test was enabled only for `walk()`. Unless you mean an even lower recursion limit?

File: Lib/os.py (hunk 2 above)
Comment (carljm) [reply Lib/os.py:555]: No, I think what you have is fine.

**Comment by carljm** [comment]
> I've enabled test_walk_above_recursion_limit for fwalk(), which tests with the recursion limit set to 45 and a directory depth of 50. Previously that test was enabled only for walk(). Unless you mean an even lower recursion limit?

No, that looks fine. Sorry, somehow I just totally missed that one-line removal when focused on the "added lines" side of the diff!

**Comment by barneygale** [comment]
No worries, thanks tons for the speedy review :)

**Pull request merged**